### ISSUE TITLE
Using alternatives as a way to dynamically avoid specific roads

### ISSUE THREAD

**Issue by rdvelazquez** [Project-OSRM/osrm-backend#5581]
__tldr:__ Is there a way to tune [these parameters](https://github.com/Azimuth1/osrm-backend/blob/master/src/engine/routing_algorithms/alternative_path_mld.cpp#L38-L172) so the routing service returns ~10 alternative routes that each differs from the primary route by a different ~10% (the 10 and 10% numbers are just rough guides wouldn't have to be exact)?

__Background:__ We would like to provide routes that dynamically avoid specific roads segments for the purpose of routing around road closures caused by flooding, downed trees, etc. for an application for first responders (fire, law enforcement and EMT). We are aware of using the traffic functionality to close specific roads but this requires rebuilding the graph which we would like to see if we can avoid for a number of reasons. One idea that we have is to have the routing service return multiple alternative routes and then select the best route that doesn't use any closed roads. We've tried tweaking [these parameters](https://github.com/Azimuth1/osrm-backend/blob/master/src/engine/routing_algorithms/alternative_path_mld.cpp#L38-L172) to get on the order of 10 alternatives but we've had limited success (some routes don't have any alternatives; some alternatives are unreasonable; parts of the route are shared by all alternatives). The best set-up we've had so far is from:
 - Upping `kSearchSpaceOverlapFactor` from 1.33 to 2
 - Increasing all instances of `kAlternativesToUnpackFactor` by 10 times
 - Increasing `kAtMostLongerBy` from 0.25 to 0.4

Any recommendations for what to try next or general thoughts on how to accomplish the goal of using alternatives to avoid specific road segments? Thanks in advance!

**Comment by danpat** [comment]
Unfortunately, the answer is "it's not that easy".  The reason that the MLD and CH graphs exist in the first place is to speed up route calculation.  They achieve this by reducing the number of candidate paths that the route search needs to traverse.  This is in direct competition with what you want for alternative route searching.  MLD doesn't prune the graph as aggressively as CH, so you will tend to get a few more alternatives out of it, but the changes are still pretty low.

There's a great (but long) paper on this topic to be found here:
http://algo2.iti.kit.edu/documents/Dissertation_Kobitzsch_Moritz.pdf

(Chapter 5 gets into the meat of current approaches for alternative path finding).

To get the flexibility you want (generate many reasonable alternatives between A->B and then select after-the-fact) would likely require a completely new approach than what's currently implemented.

**Comment by rdvelazquez** [comment]
Thanks for the quick and detailed response @danpat. Much appreciated!
